realm/realm-library/src/main/java/io/realm/internal/ColumnIndices.java

```diff
@@ -42,147 +47,91 @@
-     *
-     * @param schemaVersion the schema version
-     * @param classesMap a map of table classes to their column info
-     * @throws IllegalArgumentException if any of the ColumnInfo object is immutable.
-     */
-    public ColumnIndices(long schemaVersion, Map<Pair<Class<? extends RealmModel>, String>, ColumnInfo> classesMap) {
-        this(schemaVersion, new HashMap<>(classesMap), true);
-        for (Map.Entry<Pair<Class<? extends RealmModel>, String>, ColumnInfo> entry : classesMap.entrySet()) {
-            ColumnInfo columnInfo = entry.getValue();
-            if (mutable != columnInfo.isMutable()) {
-                throw new IllegalArgumentException("ColumnInfo mutability does not match ColumnIndices");
-            }
-            Pair<Class<? extends RealmModel>, String> classDescription = entry.getKey();
-            this.classes.put(classDescription.first, columnInfo);
-            this.classesByName.put(classDescription.second, columnInfo);
-        }
-    }
+    // Class to ColumnInfo map
+    private final Map<Class<? extends RealmModel>, ColumnInfo> classToColumnInfoMap =
+            new HashMap<Class<? extends RealmModel>, ColumnInfo>();
+    // Class name to ColumnInfo map. All the elements in this map should be existing in classToColumnInfoMap.
+    private final Map<String, ColumnInfo> simpleClassNameToColumnInfoMap =
+            new HashMap<String, ColumnInfo>();
 
-    /**
-     * Create a copy of the passed ColumnIndices with the specified mutability.
-     *
-     * @param other the ColumnIndices object to copy
-     * @param mutable if false the object is effectively final.
-     */
-    public ColumnIndices(ColumnIndices other, boolean mutable) {
-        this(other.schemaVersion, new HashMap<Pair<Class<? extends RealmModel>, String>, ColumnInfo>(other.classesToColumnInfo.size()), mutable);
-        for (Map.Entry<Pair<Class<? extends RealmModel>, String>, ColumnInfo> entry : other.classesToColumnInfo.entrySet()) {
-            ColumnInfo columnInfo = entry.getValue().copy(mutable);
-            Pair<Class<? extends RealmModel>, String> key = entry.getKey();
-            this.classes.put(key.first, columnInfo);
-            this.classesByName.put(key.second, columnInfo);
-            this.classesToColumnInfo.put(key, columnInfo);
-        }
-    }
+    private final RealmProxyMediator mediator;
+    // Due to the nature of Object Store's Realm::m_schema, SharedRealm's OsObjectSchemaInfo object is fixed after set.
+    private final OsSchemaInfo osSchemaInfo;
 
-    private ColumnIndices(long schemaVersion, Map<Pair<Class<? extends RealmModel>, String>, ColumnInfo> classesMap, boolean mutable) {
-        this.schemaVersion = schemaVersion;
-        this.classesToColumnInfo = classesMap;
-        this.mutable = mutable;
-        this.classes = new HashMap<>(classesMap.size());
-        this.classesByName = new HashMap<>(classesMap.size());
-    }
 
     /**
```

Comment:
`ColumnInfo`  is not shared between threads, but shared between `RealmObject`s for the same table and `Realm` instance where those `RealmObject`s belongs.
When an index of some field is updated, `Realm` just updates a `ColumnInfo` instance that the `Realm` is referring, and the change is seen from all `RealmObject`s because all the `RealmObject`s are referring the same instance.